### PR TITLE
arch/imxrt: initialize pwm in imxrt1020/1050 board

### PR DESCRIPTION
**Pull request by an4967**
- add init process on boot

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [89f98c85feb9ac3eff37674e2030248917017a55] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [89f98c85feb9ac3eff37674e2030248917017a55] - Code Rule Check OK.

File: os/arch/arm/src/imxrt/imxrt_pwm.c [hunk 1]
```diff
@@ -874,11 +890,39 @@ static int imxrt_pwm_setup(FAR struct pwm_lowerhalf_s *dev)
...
+	priv->base->SM[kPWM_Module_1].CTRL2 |= PWM_CTRL2_FORCE(1);
+	priv->base->SM[kPWM_Module_2].CTRL2 |= PWM_CTRL2_FORCE(1);
+
+	printf("info->duty: %d, info->frequency: %d\n", info->duty, info->frequency);
+
+	if (priv->frequency != info->frequency || priv->duty != info->duty) {
+		imxrt_pwm_stoptimer(priv->base, kPWM_Control_Module_0 | kPWM_Control_Module_1 | kPWM_Control_Module_2);
+
+		for (int i = 0; i < IMXRT_PWM_CNT; i++) {
+			imxrt_pwm_drv_init3phpwm(i, info->duty, info->frequency);
```
Comment (Taejun-Kwon) [inline os/arch/arm/src/imxrt/imxrt_pwm.c]: Logic is totally wrong.
You are controlling another path's dev here, in other words you are modifying another iNode without each iNode's file operation here.
Please make sure that you have to control each of iNode with each of fd.
Please assume that some threads using each of fd from each of diffrent path.....

File: os/arch/arm/src/imxrt/imxrt_pwm.c [hunk 2]
```diff
@@ -905,6 +949,11 @@ static int imxrt_pwm_stop(FAR struct pwm_lowerhalf_s *dev)
 {
 	FAR struct imxrt_pwmtimer_s *priv = (FAR struct imxrt_pwmtimer_s *)dev;
 
+	priv->frequency = 0;
+	priv->duty = 0;
+
+	lldbg("STOP!!!\n");
```
Comment (Taejun-Kwon) [inline os/arch/arm/src/imxrt/imxrt_pwm.c]: This is kind of verbose, please modify it.

File: os/arch/arm/src/imxrt/imxrt_pwm.c [hunk 3]
```diff
@@ -954,9 +1003,41 @@ static int imxrt_pwm_shutdown(FAR struct pwm_lowerhalf_s *dev)
...
 static int imxrt_pwm_ioctl(FAR struct pwm_lowerhalf_s *dev, int cmd,
-						 unsigned long arg)
+						   unsigned long arg)
 {
-	return -ENOTTY;
+	FAR struct imxrt_pwmtimer_s *priv = (FAR struct imxrt_pwmtimer_s *)dev;
+	int ret = -ENOTTY;
+
+	switch (cmd) {
+	case PWMIOC_IDLE:
```
Comment (Taejun-Kwon) [inline os/arch/arm/src/imxrt/imxrt_pwm.c]: IDLE is unnecessary, please read definition of start & stop in pwm.h

File: os/arch/arm/src/imxrt/imxrt_pwm.c [hunk 4]
```diff
@@ -1303,13 +1375,13 @@ FAR struct pwm_lowerhalf_s *imxrt_pwminitialize(int timer)
...
 
-		imxrt_pwm_drv_init3phpwm(timer);
+		imxrt_pwm_drv_init3phpwm(timer, DEFAULT_DUTYCYCLE, DEFAULT_FREQUENCY);
 
 		lower = (struct pwm_lowerhalf_s *)&g_pwm0_1;
 	} else
 #endif
 #ifdef CONFIG_IMXRT_PWM2
-	if (timer == 1) {
+		if (timer == 1) {
```
Comment (Taejun-Kwon) [inline os/arch/arm/src/imxrt/imxrt_pwm.c]: please align code below.

File: os/arch/arm/src/imxrt/imxrt_pwm.c [hunk 5]
```diff
@@ -86,6 +95,9 @@ struct imxrt_pwmtimer_s {
 
 	PWM_Type *base;
 	uint8_t id;
+
```
Comment (sangam-swami) [inline os/arch/arm/src/imxrt/imxrt_pwm.c]: remove new line

File: os/arch/arm/src/imxrt/imxrt_pwm.c [hunk 6]
```diff
@@ -78,6 +78,15 @@
 #include "chip/imxrt_memorymap.h"
 #include "imxrt_pwm.h"
 
+#if defined(CONFIG_ARCH_CHIP_FAMILY_IMXRT102x)
+#define IMXRT_PWM_CNT 2
+#elif defined(CONFIG_ARCH_CHIP_FAMILY_IMXRT105x)
+#define IMXRT_PWM_CNT 4
```
Comment (sunghan-chang) [inline os/arch/arm/src/imxrt/imxrt_pwm.c]: Why don't you use `PWM_CNT_COUNT`? Why do you define another definition for the same?

File: os/board/imxrt1020-evk/src/imxrt_boot.c [hunk 7]
```diff
@@ -165,6 +165,39 @@ static void imxrt_gpio_initialize(void)
 #endif
 }
 
+/****************************************************************************
+ * Name: board_pwmm_initialize
```
Comment (sunghan-chang) [inline os/board/imxrt1020-evk/src/imxrt_boot.c]: check a name

File: os/arch/arm/src/imxrt/imxrt_pwm.c (hunk 4 above)
Comment (sunghan-chang) [reply os/arch/arm/src/imxrt/imxrt_pwm.c]: You should remove this adding a tab. Same as below.

File: os/arch/arm/src/imxrt/imxrt_pwm.c [hunk 8]
```diff
@@ -986,44 +1067,48 @@ static void imxrt_pwm_reset(FAR struct pwm_lowerhalf_s *dev, int timer)
...
 static struct imxrt_pwmtimer_s g_pwm0_1 = {
-	.ops	= &g_pwm_ops,
-	.id		= 1,
-	.base	= (PWM_Type *)PWM1_BASE,
-};
+	.ops = &g_pwm_ops,
+	.id = 1,
+	.base = (PWM_Type *)PWM1_BASE,
+	.duty = 0,
+	.frequency = 0};
```
Comment (NilabhShant) [inline os/arch/arm/src/imxrt/imxrt_pwm.c]: close brace should be in new line.

File: os/arch/arm/src/imxrt/imxrt_pwm.c [hunk 9]
```diff
@@ -986,44 +1067,48 @@ static void imxrt_pwm_reset(FAR struct pwm_lowerhalf_s *dev, int timer)
...
 static struct imxrt_pwmtimer_s g_pwm0_3 = {
-	.ops	= &g_pwm_ops,
-	.id		= 3,
-	.base	= (PWM_Type *)PWM3_BASE,
-};
+	.ops = &g_pwm_ops,
+	.id = 3,
+	.base = (PWM_Type *)PWM3_BASE,
+	.duty = 0,
+	.frequency = 0};
```
Comment (NilabhShant) [inline os/arch/arm/src/imxrt/imxrt_pwm.c]: same as above.

File: os/board/imxrt1020-evk/src/imxrt_boot.c [hunk 10]
```diff
@@ -165,6 +165,39 @@ static void imxrt_gpio_initialize(void)
...
+	int ret;
+	int i;
+
+	for (i = 0; i < PWM_CNT_COUNT; i++) {
+		pwm = imxrt_pwminitialize(i);
+		if (!pwm) {
+			lldbg("Failed to get imxrt PWM lower half\n");
+			return;
+		}
+
```
Comment (NilabhShant) [inline os/board/imxrt1020-evk/src/imxrt_boot.c]: Please remove extra line.

**Comment by seinfra** [comment]
Target : [223b6e9103f5aadc6439941686de8b301c60c986] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [223b6e9103f5aadc6439941686de8b301c60c986] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [01a828c9cff9c8fa89aeda17e5b431207c25e799] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [01a828c9cff9c8fa89aeda17e5b431207c25e799] - Code Rule Check OK.

**Pull request closed without merging**